### PR TITLE
PoC: rip out the allocator API

### PR DESCRIPTION
**Pull request by erlend-aasland**

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Alternative to #31

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -3150,11 +3093,12 @@ def get_displayname(self, i: int) -> str:
...
-        add, out = text_accumulator()
-        add(f"  {self.name}\n")
+        lines = []
+        lines.append(f"  {self.name}")
         for line in self.docstring.split("\n"):
-            add(f"    {line}\n")
-        return out().rstrip()
+            lines.append(f"    {line}")
+        out = "\n".join(lines)
+        return out.rstrip()
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:3101]: I can't make it a suggested change as it spans deleted lines, but I would write the function like this:

```py
        lines = [f"  {self.name}"]
        lines.extend(f"    {line}" for line in self.docstring.split("\n"))
        return "\n".join(lines).rstrip()
```

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -6375,26 +6320,22 @@ def add_parameter(text: str) -> None:
...
     def format_docstring_parameters(params: list[Parameter]) -> str:
         """Create substitution text for {parameters}"""
-        add, output = text_accumulator()
-        for p in params:
-            if p.docstring:
-                add(p.render_docstring())
-                add('\n')
-        return output()
+        lines = [p.render_docstring() + "\n" for p in params if p.docstring]
+        return "".join(lines)
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:6338]:
```suggestion
        return "".join(p.render_docstring() + "\n" for p in params if p.docstring)
```

We could even just get rid of this method and inline the call: it's only called in one place.

File: Tools/clinic/clinic.py [hunk 3]
```diff
@@ -319,51 +278,34 @@ def linear_format(s: str, **kwargs: str) -> str:
...
-    If the last line is empty, prefix is not prepended
-    to it.  (If s is blank, returns s unchanged.)
+    If the last line is empty, it remains unchanged.
+    If s is blank, returns s unchanged.
 
     (textwrap.indent only adds to non-blank lines.)
     """
-    split = s.split('\n')
+    split = s.split("\n")
     last = split.pop()
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:298]:
```suggestion
    *split, last = s.split("\n")
```

**Review by AlexWaygood — COMMENTED** [review]
Yes!! A few suggestions to simplify the code even more. I've checked locally that all tests pass with them applied

File: Tools/clinic/clinic.py (hunk 2 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:6338]: Yes, absolutely.

Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: OTOH, the method name increases the readability slightly.

File: Tools/clinic/clinic.py [hunk 4]
```diff
@@ -294,18 +253,18 @@ def linear_format(s: str, **kwargs: str) -> str:
...
 
-    add, output = text_accumulator()
+    lines = []
     for line in s.split('\n'):
         indent, curly, trailing = line.partition('{')
         if not curly:
-            add(line)
-            add('\n')
+            lines.append(line)
+            lines.append('\n')
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:261]:
```suggestion
            lines.extend([line, "\n"])
```

File: Tools/clinic/clinic.py [hunk 5]
```diff
@@ -294,18 +253,18 @@ def linear_format(s: str, **kwargs: str) -> str:
...
             continue
 
         name, curly, trailing = trailing.partition('}')
         if not curly or name not in kwargs:
-            add(line)
-            add('\n')
+            lines.append(line)
+            lines.append('\n')
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:267]:
```suggestion
            lines.extend([line, "\n"])
```

File: Tools/clinic/clinic.py [hunk 6]
```diff
@@ -319,51 +278,33 @@ def linear_format(s: str, **kwargs: str) -> str:
         if not value:
             continue
 
-        value = textwrap.indent(rstrip_lines(value), indent)
-        add(value)
-        add('\n')
+        stripped = [line.rstrip() for line in value.split("\n")]
+        value = textwrap.indent("\n".join(stripped), indent)
+        lines.append(value)
+        lines.append('\n')
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:284]:
```suggestion
        lines.extend([value, "\n"])
```

File: Tools/clinic/clinic.py [hunk 7]
```diff
@@ -1856,14 +1796,16 @@ def render_option_group_parsing(
...
-        add("    default:\n")
+        out.append("    default:\n")
         s = '        PyErr_SetString(PyExc_TypeError, "{} requires {} to {} arguments");\n'
-        add(s.format(f.full_name, count_min, count_max))
-        add('        goto exit;\n')
-        add("}")
-        template_dict['option_group_parsing'] = format_escape(output())
+        out.append(s.format(f.full_name, count_min, count_max))
+        out.append('        goto exit;\n')
+        out.append("}")
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:1805]:
```suggestion
        out.extend([
            s.format(f.full_name, count_min, count_max),
            '        goto exit;\n',
            '}'
        ])
```

File: Tools/clinic/clinic.py [hunk 8]
```diff
@@ -6293,12 +6232,12 @@ def add_parameter(text: str) -> None:
                 else:
                     s = ' ' + text
                     if line_length + len(s) >= 72:
-                        add('\n')
-                        add(indent)
+                        lines.append('\n')
+                        lines.append(indent)
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:6236]:
```suggestion
                        lines.extend(['\n', indent])
```

File: Tools/clinic/clinic.py [hunk 9]
```diff
@@ -6321,8 +6260,8 @@ def add_parameter(text: str) -> None:
                     added_star = True
                     add_parameter('*,')
 
-                p_add, p_output = text_accumulator()
-                p_add(fix_right_bracket_count(p.right_bracket_count))
+                p_lines = []
+                p_lines.append(fix_right_bracket_count(p.right_bracket_count))
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:6264]:
```suggestion
                p_lines = [fix_right_bracket_count(p.right_bracket_count)]
```

File: Tools/clinic/clinic.py (hunk 2 above)
Comment (AlexWaygood) [reply Tools/clinic/clinic.py]: Sure, we can keep it as a separate method!

**Comment by erlend-aasland** [comment]
Moved upstream!

- python/cpython#113402

**Comment by erlend-aasland** [comment]
(I'll incorporate your latest suggestions over there)

File: Tools/clinic/clinic.py (hunk 7 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:1805]: Hm, I prefer the existing variant.

Comment (AlexWaygood) [reply Tools/clinic/clinic.py:1805]: Fair enough, I don't feel strongly! The existing variant is fine

**Pull request closed without merging**